### PR TITLE
Export variants of image sizes and file attachments

### PR DESCRIPTION
**Pull request by ChrisBAshton**
When we import content from Whitehall, we need to ensure that we
transfer all the variants of image sizes and attachments too, in case they are
used by frontend applications. This is particularly true of image thumbnails.
There is no easy way to query these variants from outside of Whitehall, so this
work is best done as part of the export.

Trello card: https://trello.com/c/XOJ64DPo/1151-export-variants-of-images-and-attachments

Screenshot of new data structure

### CODE REVIEW

File: app/models/image.rb [hunk 1]
```diff
@@ -15,6 +15,14 @@ def url(*args)
     image_data.file_url(*args)
   end
 
+  def variants
+    sizes = {}
+    image_data.file.versions.each do |variant, details|
+      sizes[variant] = details.url
+    end
+    sizes
```
Comment (kevindew) [inline app/models/image.rb]: Minor, you might want to look at using each_with_object for this.

File: app/models/image.rb [hunk 2]
```diff
@@ -15,6 +15,14 @@ def url(*args)
     image_data.file_url(*args)
   end
 
+  def variants
+    sizes = {}
+    image_data.file.versions.each do |variant, details|
+      sizes[variant] = details.url
```
Comment (kevindew) [inline app/models/image.rb]: Nice one digging in and finding this out 👍

File: test/unit/presenters/document_export_presenter_test.rb [hunk 3]
```diff
@@ -128,6 +148,25 @@ class DocumentExportPresenterTest < ActiveSupport::TestCase
     assert_equal "FileAttachment", attachment[:type]
     assert_equal publication_file.attachments.first.attachment_data.as_json.symbolize_keys,
                  attachment[:attachment_data]
+    # TODO - do titles vary per attachment variant?
+    assert_match(/^file\-attachment\-title\-[0-9]$/, attachment[:title])
```
Comment (ChrisBAshton) [inline test/unit/presenters/document_export_presenter_test.rb]: I'd like to double check that we don't expect each variant of FileAttachment to have its own title (e.g. "File attachment PDF" vs "File attachment thumbnail").

Currently I believe we just have `:title` on the FileAttachment root, which in this hypothetical scenario would be "File attachment".

File: test/unit/presenters/document_export_presenter_test.rb [hunk 4]
```diff
@@ -145,6 +145,7 @@ class DocumentExportPresenterTest < ActiveSupport::TestCase
     attachment = result.dig(:editions, 0, :attachments, 0)
 
     assert_equal publication_file.attachments.first.url, attachment[:url]
+    assert_match(/^file\-attachment\-title\-[0-9]$/, attachment[:title])
```
Comment (ChrisBAshton) [inline test/unit/presenters/document_export_presenter_test.rb]: I'd like to double check that we don't expect each variant of FileAttachment to have its own title (e.g. "File attachment PDF" vs "File attachment thumbnail").

Currently I believe we just have `:title` on the FileAttachment root, which in this hypothetical scenario would be "File attachment".

File: app/presenters/document_export_presenter.rb [hunk 5]
```diff
@@ -68,6 +68,18 @@ def present_attachments(edition)
                          end
       attachment.as_json(include: :attachment_data, methods: %i[url type])
                 .merge(govspeak_content)
+                .merge(variants: attachment_variants(attachment))
+    end
+  end
+
+  def attachment_variants(attachment)
+    return [] unless attachment.try(:attachment_data)
```
Comment (kevindew) [inline app/presenters/document_export_presenter.rb]: Should this return a hash to be consistent with the next statement?

Comment (ChrisBAshton) [reply app/presenters/document_export_presenter.rb]: Whoops - good spot - it should!

**Pull request merged**